acvm-repo/acir/src/circuit/opcodes.rs

```diff
@@ -146,16 +146,170 @@ pub enum Opcode<F: AcirField> {
+                write!(f, "EXPR ")?;
+
+                // Check if it's `w - C = 0` and print it as `w = C`,
+                // or if its' `w + C = 0` and print it as `w = -C`.
+                if expr.mul_terms.is_empty() && expr.linear_combinations.len() == 1 {
+                    let (coefficient, witness) = expr.linear_combinations[0];
+                    let constant = expr.q_c;
+
+                    if coefficient.is_one() {
+                        // This is `wN + C = 0`, so show it as `wN = -C`.
+                        let constant = -constant;
+                        write!(f, "{witness} = {constant}")?;
+                        return Ok(());
+                    }
+
+                    if (-coefficient).is_one() {
+                        // This is `-wN + C = 0`, so show it as `wN = C`.
+                        write!(f, "{witness} = {constant}")?;
+                        return Ok(());
+                    }
+                }
+
+                // Check if it's `-w1 * w2 = 0` or `w1 * -w2 = 0` in which case we can
+                // print it as `w1 = w2`.
+                if expr.mul_terms.is_empty()
+                    && expr.linear_combinations.len() == 2
+                    && expr.q_c.is_zero()
+                {
+                    let (coefficient1, witness1) = expr.linear_combinations[0];
+                    let (coefficient2, witness2) = expr.linear_combinations[1];
+
+                    if (coefficient1.is_one() && (-coefficient2).is_one())
+                        || ((-coefficient1).is_one() && coefficient2.is_one())
+                    {
+                        write!(f, "{witness1} = {witness2}")?;
+                        return Ok(());
+                    }
+                }
+
+                // Check if it's `c1*w1*w2 - w3 = 0` or similar, in which case we can print it
+                // as `w1 = c1*w2*w3`.
+                if expr.mul_terms.len() == 1
+                    && expr.linear_combinations.len() == 1
+                    && expr.q_c.is_zero()
+                {
+                    let (coefficient1, witness1, witness2) = expr.mul_terms[0];
+                    let (coefficient2, witness3) = expr.linear_combinations[0];
+                    if witness1 != witness3
+                        && witness2 != witness3
+                        && (coefficient2.is_one() || (-coefficient2).is_one())
+                    {
+                        let coefficient1 =
+                            if coefficient2.is_one() { -coefficient1 } else { coefficient1 };
+                        if coefficient1.is_one() {
+                            write!(f, "{witness3} = {witness1}*{witness2}")?;
+                        } else if (-coefficient1).is_one() {
+                            write!(f, "{witness3} = -{witness1}*{witness2}")?;
+                        } else {
+                            write!(f, "{witness3} = {coefficient1}*{witness1}*{witness2}")?;
+                        }
```

Comment:
Just my personal opinion, but I think this rule is going one step too far. There are 3 witnesses, so it's not clear that one form is better than the other. It reminds me of a quadratic equation, which I can't remember being reorganised depending on whether there is a non-zero constant or not. 

On a more practical note, the others were easy to review: `w1 = w2` if the coefficients are 1 and -1. Here `coefficient1` is changing sign based on whether `coefficient2` is 1, and _then_ we look at whether the swapped sign `coefficient1` is 1 or -1. It's harder to think through whether this is correct than what it should be in a display function, it's not obvious.

Reply:
Or another potential idea to limit the amount of special casing is to only look for these two opportunities:
1. There are only 2 linear terms and nothing else, with 1 and -1, ie. `w2 = w1`. 
2. There is only 1 linear term, which can be written as `w2 = ...`, and then the RHS can be anything, e.g. `w2 = 3`, or `w2 = w1*w3 - 3*w3*w5 + 10`, the point is that it looks like we are defining the witness on the LHS, which is familiar to programmers. 

Then we don't need to special case `w3 = w1*w2` and `w1 = 5` any more, as they are covered by uniquely defining a single witness.

Reply:
I ended up doing something like this, but not exactly this. I'll write it in the general comments section.